### PR TITLE
Consolidate solution configuration xml parsing logic

### LINKED ISSUE #6751: Reduce code clones for parsing the SolutionConfiguration xml
There's multiple code clones that either parse or emit the xml for the `CurrentSolutionConfigurationContents` xml (see details https://github.com/dotnet/msbuild/pull/6738) that VS sets as a global property: 
- Microsoft.Build.Tasks.ResolveProjectBase.CacheProjectElementsFromXml
- Microsoft.Build.Construction.SolutionProjectGenerator.AddPropertyGroupForSolutionConfiguration
- Microsoft.Build.Experimental.ProjectCache.ProjectCacheService.PostCacheRequest.GenerateGraphEntryPointsFromSolutionConfigurationXml

To increase maintainability and reuse it would be nice to create an object model for this structure which can serialize / deserialize itself to / from the xml.

1st option: Add new classes. Unfortunately there will be a bit of an overlap between these and the existing Microsoft.Build.Construction.SolutionFile, Microsoft.Build.BuildEngine.Shared.ProjectInSolution, and Microsoft.Build.BuildEngine.Shared.ProjectConfigurationInSolution

2nd option: Massage the existing SolutionFile, ProjectInSolution, and ProjectConfigurationInSolution to accomodate `CurrentSolutionConfigurationContents` as well. Unfortunately these types are already a bit messy so messing them up even further might be too much.

Regardless of which option we choose, we'll have to put the OM in shared sources because both the tasks dll and the engine dll need to depend on it.

Given these complexities and the fact that the project cache VS workaround should be deleted at some point, maybe it's not worth doing any of this. Or maybe it's worth doing a rewrite / redesign of the whole solution object model and metaproject generation logic. That area of the code is a bit of a quagmire

### PR DESCRIPTION
**Pull request by dfederm**
Fixes #6751

This change combines the solution configuration xml parsing which the project caching code and the tasks code uses. I did not add `SolutionProjectGenerator.AddPropertyGroupForSolutionConfiguration` into the mix as that's a write scenario and the other two are read scenarios. Maybe at some point the object model could be made more robust and support both the read and write scenarios, but whatever, this is incrementally better.

Note that the main motivation here is that the graph construction code currently does not support the sln-defined configurations, ie what the `AssignProjectConfiguration` target does, so I'll be working on a follow-up change for that. And instead of creating a third copy of this parsing logic, I thought I'd send this PR to consolidate the logic first.

Basically, `SolutionConfiguration.cs` is just ripped directly from `ResolveProjectBase.cs`, and then `ProjectCacheService` just uses `SolutionConfiguration` instead of parsing itself.

### CODE REVIEW

File: src/Shared/SolutionConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        public SolutionConfiguration(string xmlString)
+        {
+            XmlDocument? doc = null;
+
+            if (!string.IsNullOrEmpty(xmlString))
+            {
+                doc = new XmlDocument();
+                var settings = new XmlReaderSettings { DtdProcessing = DtdProcessing.Ignore };
+                using (XmlReader reader = XmlReader.Create(new StringReader(xmlString), settings))
```
Comment (Forgind) [inline src/Shared/SolutionConfiguration.cs:41]: Why does it need XmlReader.Create(new StringReader(xmlString), settings), whereas it looks like the previous version just called LoadXml on the string directly in ProjectCacheService? Is it because you wanted to pass DtdProcessing.Ignore? I seem to remember that's a security thing, so perhaps that was the motivation?

File: src/Shared/SolutionConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            // <SolutionConfiguration>
+            //  <ProjectConfiguration Project="{786E302A-96CE-43DC-B640-D6B6CC9BF6C0}" AbsolutePath="c:foo\Project1\A.csproj" BuildProjectInSolution="True">Debug|AnyCPU</ProjectConfiguration>
+            //  <ProjectConfiguration Project="{881C1674-4ECA-451D-85B6-D7C59B7F16FA}" AbsolutePath="c:foo\Project2\B.csproj" BuildProjectInSolution="True">Debug|AnyCPU<ProjectDependency Project="{4A727FF8-65F2-401E-95AD-7C8BBFBE3167}" /></ProjectConfiguration>
+            //  <ProjectConfiguration Project="{4A727FF8-65F2-401E-95AD-7C8BBFBE3167}" AbsolutePath="c:foo\Project3\C.csproj" BuildProjectInSolution="True">Debug|AnyCPU</ProjectConfiguration>
+            // </SolutionConfiguration>
+            //
+            if (doc?.DocumentElement != null)
+            {
+                foreach (XmlElement xmlElement in doc.DocumentElement.ChildNodes)
+                {
```
Comment (Forgind) [inline src/Shared/SolutionConfiguration.cs:58]: This step is just to cache some stuff early? Is it always needed? I'm wondering if we could do it lazily. In particular, it looks like ProjectCacheService wasn't doing all of this before.

(Partially inspired by https://github.com/dotnet/msbuild/pull/8526, which got a big improvement by avoiding GetFileInfoNoThrow; I'm wondering if GetFullPathNoThrow might have a similar problem, though we don't actually have to crack open the file, so maybe not.)

File: src/Shared/SolutionConfiguration.cs (hunk 1 above)
Comment (dfederm) [reply src/Shared/SolutionConfiguration.cs:41]: I consolidated on the behavior in `ResolveProjectBase` since it's more battle-tested.

File: src/Shared/SolutionConfiguration.cs (hunk 2 above)
Comment (dfederm) [reply src/Shared/SolutionConfiguration.cs:58]: Yes, it's so that the file can be parsed just once. I'm not sure where laziness would add much since the `SolutionConfiguration` shouldn't be created at all if it won't be used.

Comment (Forgind) [reply src/Shared/SolutionConfiguration.cs:58]: I guess where I got confused is that the previous ProjectCacheService version didn't do this early caching of parsed information step, yet it still looks like it just used fast access methods. Are things like `node.Attributes!["BuildProjectInSolution"]` secretly a slow linear search?

Comment (dfederm) [reply src/Shared/SolutionConfiguration.cs:58]: It didn't do this caching, but as this is a 1-time thing for initialization, I'm not sure it really matters too much in practice.

I'll expose this part as a static method to avoid the extra allocation while still getting the code sharing.

**Review by Forgind — APPROVED** [review]
I think I'm happy, thanks!
